### PR TITLE
refactor upgradeConfig.js

### PR DESCRIPTION
**Pull request by noahlemen**
refactored this file to be (in my opinion) a bit easier to follow

not attached to this approach, very open to other suggestions too!!

### CODE REVIEW

File: __tests__/upgradeConfig.test.js [hunk 1]
```diff
@@ -71,3 +71,14 @@ test('does not another flow preset if already present and hasFlow option passed'
   expect(upgradeConfig(config, { hasFlow: true })).toMatchSnapshot();
 });
 
+test('does not another flow preset if already present in an array and hasFlow option passed', () => {
```
Comment (kedromelon) [inline __tests__/upgradeConfig.test.js]: noticed this mid-refactor, not sure if this is a case worth handling -- but i figured it couldn't hurt to be safe

File: src/upgradeConfig.js [hunk 2]
```diff
@@ -1,110 +1,91 @@
...
-  let plugins = config.plugins;
+function appendFlowIfNeeded(presets) {
+  return presets.find(preset => preset === '@babel/preset-flow' || preset[0] === '@babel/preset-flow')
+    ? presets
+    : presets.concat(['@babel/preset-flow']);
+}
+
+function transformPreset(preset) {
+  let presetName = Array.isArray(preset) ? preset[0] : preset;
+  const presetOptions = Array.isArray(preset) ? preset[1] : {};
```
Comment (kedromelon) [inline src/upgradeConfig.js]: normalizing presets to be "__Name" and "__Options" and then transforming them back down to a string at the end allows more code to be shared preset/plugin formats

might seem a bit unnecessary right now, but it also simplifies operations on options a bit more

File: src/upgradeConfig.js [hunk 3]
```diff
@@ -1,110 +1,91 @@
...
+
+  return null;
 }
 
 module.exports = function upgradeConfig(config, options) {
   config = Object.assign({}, config);
 
-  changePresets(config, options);
-  changePlugins(config);
+  if (config.presets) {
```
Comment (kedromelon) [inline src/upgradeConfig.js]: i was thinking it might be clearer to only mutate config within `upgradeConfig`, but that might just be personal preference

**Comment by kedromelon** [comment]
dug in more on reducing duplicate code on my last commit (https://github.com/babel/babel-upgrade/pull/50/commits/6fcdef56352cc3ba66cfcd2b24297f7e727b73ea)

may have gone too far with it, but curious to see if anyone thinks parts of it make sense to use here

File: src/upgradeConfig.js [hunk 4]
```diff
@@ -1,110 +1,81 @@
...
+
+  return collection;
+}
+
+function parseConfigItem(rawConfigItem, configItemType) {
+  let name = Array.isArray(rawConfigItem) ? rawConfigItem[0] : rawConfigItem;
 
-  if (!Array.isArray(presets) && typeof presets === 'string') {
-    presets = config.presets = config.presets.split(',').map((preset) => preset.trim());
+  if (name.indexOf(`babel-${configItemType}`) !== 0 && name.indexOf('@babel/') !== 0) {
```
Comment (aulisius) [inline src/upgradeConfig.js]: Can't we use `.includes` instead of `.indexOf`? Gives a clearer intent imo.

File: src/upgradeConfig.js [hunk 5]
```diff
@@ -1,110 +1,81 @@
...
 
-  changePresets(config, options);
-  changePlugins(config);
+  upgradeConfigForEnv(config, options);
 
   if (config.env) {
-    Object.keys(config.env).forEach((env) => {
-      changePresets(config.env[env]);
-      changePlugins(config.env[env]);
+    Object.keys(config.env).forEach(env => {
```
Comment (aulisius) [inline src/upgradeConfig.js]: `Object.values` is what you'd want to use here.

File: __tests__/upgradeConfig.test.js (hunk 1 above)
Comment (aulisius) [reply __tests__/upgradeConfig.test.js]: minor grammar fix: 
```diff
- does not another
+ does not add another
```

File: src/upgradeConfig.js (hunk 5 above)
Comment (kedromelon) [reply src/upgradeConfig.js]: heh, great point. that totally didn't occur to me while moving this code around 😆

File: src/upgradeConfig.js (hunk 4 above)
Comment (kedromelon) [reply src/upgradeConfig.js]: hmm -- i agree, but i do think we _specically_ want to check the beginning of the string, and `includes` would return true if it were anywhere in it (which would probably be a typo on the user's part, but still)

Comment (kedromelon) [reply src/upgradeConfig.js]: we could use `startsWith` though? may need a polyfill depending on versions of node we're trying to support?

File: __tests__/upgradeConfig.test.js (hunk 1 above)
Comment (kedromelon) [reply __tests__/upgradeConfig.test.js]: thanks! nice catch

File: src/upgradeConfig.js (hunk 4 above)
Comment (kedromelon) [reply src/upgradeConfig.js]: ah, looks like that should be supported even in node 4. let's use `startsWith` here

File: src/upgradeConfig.js [hunk 6]
```diff
@@ -1,111 +1,78 @@
...
-  changePresets(config, options);
-  changePlugins(config);
+  upgradeConfigForEnv(config, options);
 
   if (config.env) {
-    Object.keys(config.env).forEach((env) => {
-      changePresets(config.env[env]);
-      changePlugins(config.env[env]);
-    });
+    Object.values(config.env).forEach(envConfig => upgradeConfigForEnv(envConfig, options));
```
Comment (existentialism) [inline src/upgradeConfig.js]: If we're supporting [node 4](https://github.com/babel/babel-upgrade/pull/16) we'll need to polyfill this (or just go back to Object.keys)?

Comment (kedromelon) [reply src/upgradeConfig.js]: good point -- i forgot that node 4 doesn't support this. should we be concerned that CI for node 4 isn't failing?

Comment (kedromelon) [reply src/upgradeConfig.js]: now that i look back at the tests, env tests could probably be significantly more extensive. this seems like something that should be caught by travis

Comment (existentialism) [reply src/upgradeConfig.js]: I don't think anyone will disagree that we need more tests :)

Comment (kedromelon) [reply src/upgradeConfig.js]: noticed the linked PR does add preset-env with target as node 4 as well as adds `@babel/polyfill` to `bin/babel-upgrade`, so we should be ok to use Object.values here? confirmed that it works running in node 4 for me locally

Comment (existentialism) [reply src/upgradeConfig.js]: Ah yes, that's fine then

File: src/packageData.js [hunk 7]
```diff
@@ -245,9 +245,17 @@ const packages = Object.assign(
...
+  'babel-preset-es2017',
+  'babel-preset-latest'
+]);
+
 module.exports = {
   packages,
   presets,
   plugins,
   latestPackages,
+  packagesWarnAboutEnv
```
Comment (existentialism) [inline src/packageData.js]: Where is this used?

**Comment by existentialism** [comment]
fwiw most of this code was just waiting on some more time for the tool to bake a bit, and I wouldn't have said it was high on the list of things that _needed_ immediate cleanup.

(@hzoo and I had even discussed going to an internal set of plugins that would get passed config/deps down a chain to modify, but said we'd wait and see, and keep it simple if anything)

That being said, changes look good. Thoughts @hzoo?

File: src/packageData.js (hunk 7 above)
Comment (kedromelon) [reply src/packageData.js]: hm not sure what this is, I suspect I pushed it up by accident

**Comment by kedromelon** [comment]
@existentialism that makes sense -- I completely understand if it feels too early to merge in something like this 

out of curiosity, what parts of the existing code would you prioritize higher for immediate cleanup?

File: src/packageData.js (hunk 7 above)
Comment (kedromelon) [reply src/packageData.js]: ah yeah i had started looking at the `Log when replacing out preset-es2015,16,17,latest as FYI` mentioned in the readme at some point and then accidentally added these changes to my commit

**Pull request closed without merging**